packages/alfa-dom/src/node/element.ts

```diff
@@ -447,6 +448,59 @@ export namespace Element {
     });
   }
 
+  export function cloneElement(
+    element: Element,
+    newElements: Element[],
+    predicate: Predicate<Element>,
+    device?: Device,
+  ): Trampoline<Element> {
+    return Trampoline.traverse(element.children(), (child) => {
+      if (Element.isElement(child) && predicate(child)) {
+        return Trampoline.done(newElements);
+      }
+
+      return Node.cloneNode(child, newElements, predicate, device).map(
+        (node) => [node],
+      );
+    }).map((children) => {
+      const deviceOption = Option.from(device);
+      const clonedElement = Element.of(
+        element.namespace,
+        element.prefix,
+        element.name,
+        element.attributes.map((attribute) =>
+          Attribute.cloneAttribute(attribute).run(),
+        ),
+        Iterable.flatten(children),
+        element.style.map((block) => {
+          return Block.of(
+            Iterable.map(block.declarations, (declaration) =>
+              Declaration.of(
+                declaration.name,
+                declaration.value,
+                declaration.important,
+              ),
+            ),
+          );
+        }),
+        deviceOption.flatMap((d) => element.getBoundingBox(d)),
+        deviceOption,
+        element.externalId,
+        element.extraData,
```

Comment:
**Issue:** Also realising that this will actually share the same `extraData` object, which might create some confusion.
```typescript
const foo = h.element(…, { extraStuff: "foo" });

const bar = Node.clone(foo);
bar.extraData.extraStuff = "bar";
```
This also change `foo.extraData.ExtraStuff`.

On on side, it is not that bad because Alfa doesn't care, but it feels like this can still create confusion in users. We could:
* Keep it that way. Make it clear that `extraData` is copied with referencial equality, with all that it implies.
* Discard `extraData` when cloning, make it clear that it is the behaviour.
* Offer some API to pass an `extraData` cloning function.

The third option seems a bit crazy. Given that we do not currently have any actual use for `extraData`, I'd rather go for the second one. We can revisit later if needed.

---
Also, we should make it clear in the function's documentation that `externalId` is going to be copied to the clone. This is likely the desired behavior (at least it is for our use case), but it could be confusing if undocumented.

Reply:
I went with the second suggestion and discarded the `extraData`.